File: src/routes/comments.py
from flask import Blueprint, request, jsonify
from src.models.user import db, Comment, CommentLike
from src.routes.auth import require_auth

comments_bp = Blueprint('comments', __name__)

@comments_bp.route('/comments/<int:comment_id>/like', methods=['POST'])
@require_auth
def toggle_comment_like(comment_id):
    try:
        comment = Comment.query.get_or_404(comment_id)
        user_id = request.current_user.id

        # 檢查是否已經點讚
        existing_like = CommentLike.query.filter_by(comment_id=comment_id, user_id=user_id).first()

        if existing_like:
            # 取消點讚
            db.session.delete(existing_like)
            comment.likes_count = max(0, comment.likes_count - 1)
            action = 'unliked'
        else:
            # 添加點讚
            like = CommentLike(comment_id=comment_id, user_id=user_id)
            db.session.add(like)
            comment.likes_count += 1
            action = 'liked'

        db.session.commit()

        return jsonify({
            'message': f'評論{action}',
            'likes_count': comment.likes_count,
            'liked': action == 'liked'
        }), 200

    except Exception as e:
        db.session.rollback()
        return jsonify({'error': f'點讚操作失敗: {str(e)}'}), 500

@comments_bp.route('/comments/<int:comment_id>', methods=['DELETE'])
@require_auth
def delete_comment(comment_id):
    try:
        comment = Comment.query.get_or_404(comment_id)
        
        # 檢查是否是評論作者
        if comment.user_id != request.current_user.id:
            return jsonify({'error': '只能刪除自己的評論'}), 403

        # 更新帖子的評論數
        post = comment.post
        post.comments_count = max(0, post.comments_count - 1)

        db.session.delete(comment)
        db.session.commit()

        return jsonify({'message': '評論已刪除'}), 200

    except Exception as e:
        db.session.rollback()
        return jsonify({'error': f'刪除評論失敗: {str(e)}'}), 500

File: src/routes/comments.ts
import { Hono } from 'hono';
import { HTTPException } from 'hono/http-exception';
import { requireAdmin } from './auth';

export const commentRoutes = new Hono<{ Bindings: Env }>();

// Like/unlike a comment
commentRoutes.post('/:id/like', async (c) => {
  const commentId = parseInt(c.req.param('id'));
  const user = c.get('user');
  const db = c.env.DB;

  const comment = await db.prepare('SELECT * FROM comments WHERE id = ?').bind(commentId).first();
  if (!comment) {
    throw new HTTPException(404, { message: 'Comment not found' });
  }

  const existingLike = await db.prepare('SELECT * FROM comment_likes WHERE comment_id = ? AND user_id = ?')
    .bind(commentId, user.id).first();

  if (existingLike) {
    // Unlike
    await db.prepare('DELETE FROM comment_likes WHERE comment_id = ? AND user_id = ?').bind(commentId, user.id).run();
    await db.prepare('UPDATE comments SET likes_count = likes_count - 1 WHERE id = ?').bind(commentId).run();
    return c.json({ message: 'Comment unliked', liked: false });
  } else {
    // Like
    await db.prepare('INSERT INTO comment_likes (comment_id, user_id) VALUES (?, ?)').bind(commentId, user.id).run();
    await db.prepare('UPDATE comments SET likes_count = likes_count + 1 WHERE id = ?').bind(commentId).run();
    return c.json({ message: 'Comment liked', liked: true });
  }
});

// Delete a comment (admin only)
commentRoutes.delete('/:id', requireAdmin, async (c) => {
  const commentId = parseInt(c.req.param('id'));
  const db = c.env.DB;

  const comment = await db.prepare('SELECT * FROM comments WHERE id = ?').bind(commentId).first();
  if (!comment) {
    throw new HTTPException(404, { message: 'Comment not found' });
  }

  // Delete related likes first
  await db.prepare('DELETE FROM comment_likes WHERE comment_id = ?').bind(commentId).run();
  
  const { success } = await db.prepare('DELETE FROM comments WHERE id = ?').bind(commentId).run();

  if (!success) {
    throw new HTTPException(500, { message: 'Failed to delete comment' });
  }

  // Update post comment count
  await db.prepare('UPDATE posts SET comments_count = comments_count - 1 WHERE id = ?').bind(comment.post_id).run();

  return c.json({ message: 'Comment deleted successfully' });
});

